### ISSUE TITLE
Mapping sequence logit to amino acid identity for ESM3 and ESMC

### ISSUE THREAD

**Issue by kentwait** [evolutionaryscale/esm#252]
First, I know this has been asked in #82 and #86 but I does not explicitly address my question so I am opening a new issue.

I would like to explicitly clarify how to read the 64-dim sequence logits output in the `LogitsOutput` object.

I am porting some code from ESM2 to work with ESM3 and ESMC. I am using esm 3.2.0.
Looking at the ESM3 code, `SEQUENCE_VOCAB` in `esm/esm/utils/constants/esm3.py` has 33 items in its list - the same set and order of tokens in ESM2.
```
SEQUENCE_VOCAB = [
    "<cls>", "<pad>", "<eos>", "<unk>",
    "L", "A", "G", "V", "S", "E", "R", "T", "I", "D", "P", "K",
    "Q", "N", "F", "Y", "M", "H", "W", "C", "X", "B", "U", "Z",
    "O", ".", "-", "|",
    "<mask>",
]
```
For example, in ESM2 index 0 of the logits tensor corresponds to the logit for the <cls> (BOS) token. Or index 5 for L, etc.

As mentioned in section A.1.3

> Protein sequences are tokenized as the 20 canonical amino acids, plus BOS, EOS, mask, pad, unknown. 
We keep four non-standard amino acids as in Lin et al. (5), B - Asparagine, U - selenocysteine, Z - glutamic acid, and O - ornithine. This totals to 29 tokens.

Does this mean that indices 0-23, 25-28, and 32  in the 64-dim tensor correspond to `["<cls>", "<pad>", "<eos>", "<unk>", "L", "A", "G", "V", "S", "E", "R", "T", "I", "D", "P", "K", "Q", "N", "F", "Y", "M", "H", "W", "C", "B", "U", "Z", "O", "<mask>"]`? Or conversely, indices 24, 29-31 and 33-63 of the 64-dim tensor should be ignored? 

Or is there a different ordering such that only the first 29 of the 64 values are meaningful?

---

For example, I have a sequence `KA<mask>ISQ` masked at index 2 and I want to find what the model thinks is the most likely amino acid at that masked position.

```
from esm.models.esm3 import ESM3
from esm.sdk.api import ESMProtein, LogitsConfig

model = ESM3.from_pretrained("esm3_sm_open_v1").to("cuda")
protein = ESMProtein(sequence="KA<mask>ISQ")
protein_tensor = model.encode(protein)
print(protein_tensor)
```

The model's tokenizer converts this to `[ 0, 15,  5, 32, 12,  8, 16,  2]`, inserting a <cls> (BOS) and <eos> tokens at both ends respectively.

```
ESMProteinTensor(sequence=tensor([ 0, 15,  5, 32, 12,  8, 16,  2], device='cuda:0'), structure=None, secondary_structure=None, sasa=None, function=None, residue_annotations=None, coordinates=None, potential_sequence_of_concern=False)
``` 

Running the model using `.logits` method outputs a `LogitsOutput` object that contains a `ForwardTrackData` for the logits that was requested (in this case the sequence logits).

```
logits_output = model.logits(
   protein_tensor, LogitsConfig(sequence=True, return_embeddings=True)
)
print(logits_output.logits.sequence.shape)  # torch.Size([1, 8, 64])
print(logits_output.logits.sequence[0, 3])  # prints logits for the <mask> token in the input KA<mask>ISQ
```

The output is a 64-dim tensor below, the vocabulary in the code has a length of 33, but the section A.1.3 specifies only 29 tokens are used. I want to know the corresponding logit for each token but the index mapping is unclear. 

```
tensor([-1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01,  3.3008e-01,
         1.1768e-01, -1.4648e-01, -8.8867e-02,  2.4121e-01, -1.5137e-01,
        -7.2754e-02, -9.1553e-03, -1.1670e-01, -4.4141e-01, -3.0469e-01,
         9.8877e-03, -4.7656e-01, -4.0430e-01, -5.1953e-01, -7.8516e-01,
        -1.2891e+00, -1.0078e+00, -1.5312e+00, -1.3672e+00, -3.7344e+00,
        -1.3938e+01, -1.4000e+01, -1.5500e+01, -1.8500e+01, -1.8625e+01,
        -1.8625e+01, -1.8625e+01, -1.8750e+01, -1.8625e+01, -1.8625e+01,
        -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01,
        -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8750e+01, -1.8750e+01,
        -1.8625e+01, -1.8625e+01, -1.8750e+01, -1.8750e+01, -1.8625e+01,
        -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01,
        -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01,
        -1.8625e+01, -1.8625e+01, -1.8625e+01, -1.8625e+01], device='cuda:0')
```

Also, in this tensor it appears that values less than index 4 and greater than 28 are meaningless (-1.8625e+01). At first I thought this was useful information, but that's only 25 values but there should be 29 tokens utilized in the vocabulary.
I hope you can clarify this matter.

**Comment by bschilder** [comment]
@kentwait I made an `Alphabet` class to make this more intuitive. 
It match the esm1/2 syntax a bit closer too.

@yxliu-TAMU let me know if you spot anything that's incorrect.

```py
def get_sequence_tokenizer():
    """
    Get the sequence tokenizer for the ESM3 model.
    
    The output logits are 1 x seq_len x 64: this is not the true alphabet size (33).
    See:
        https://github.com/evolutionaryscale/esm/issues/86
        https://github.com/evolutionaryscale/esm/issues/252
    """
    from esm.tokenization.sequence_tokenizer import EsmSequenceTokenizer 
    Tokenizer = EsmSequenceTokenizer() 
    return Tokenizer

def decode_token(token_id,
                 tokenizer=None):
    """
    Decode a token ID to a string.

    The output logits are 1 x seq_len x 64: this is not the true alphabet size (33).
    See:
        https://github.com/evolutionaryscale/esm/issues/86
        https://github.com/evolutionaryscale/esm/issues/252
    """
    if tokenizer is None:
        tokenizer = get_sequence_tokenizer()
    return tokenizer.decode(token_id)


class Alphabet:
    def __init__(self, tokenizer=None):
        if tokenizer is None:
            tokenizer = get_sequence_tokenizer()
        token_id_to_str = {i: tokenizer.decode(i) for i in range(64)}
        str_to_token_id = {v: k for k, v in token_id_to_str.items()}
        self.tokenizer = tokenizer
        self.str_to_token_id = str_to_token_id
        self.token_id_to_str = token_id_to_str

    def get_token(self, idx):
        if not isinstance(idx, int):
            raise ValueError(f"Index must be an integer, got {type(idx)}")
        return self.token_id_to_str[idx]

    def get_idx(self, token):
        if not isinstance(token, str):
            raise ValueError(f"Token must be a string, got {type(token)}")
        return self.str_to_token_id[token]
    
    def __len__(self):
        return len(self.tokenizer.all_token_ids)
    
    def __len_full__(self):
        return len(self.token_id_to_str)
```

**Comment by imathur1** [comment]
ESM3/ESMC weren't supervised to predict special tokens `<cls>`, `<pad>`, `<eos>`, `<unk>` (vocab index <4) or `.`, `-`, `|`, `<mask>` (vocab index >28) which is why they're meaningless.

To your question about the index mapping, indices 0 to 32 out of 63 correspond to the tokens in the vocab, and the rest have no meaning!
